File: src/main/java/frc/robot/RobotContainer.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package frc.robot;

import static edu.wpi.first.units.Units.*;

import com.ctre.phoenix6.hardware.Pigeon2;
import com.ctre.phoenix6.hardware.TalonFX;
import com.ctre.phoenix6.swerve.SwerveModule.DriveRequestType;
import com.ctre.phoenix6.swerve.SwerveRequest;
import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
import edu.wpi.first.wpilibj2.command.Command;
import edu.wpi.first.wpilibj2.command.Commands;
import edu.wpi.first.wpilibj2.command.InstantCommand;
import edu.wpi.first.wpilibj2.command.WaitCommand;
import edu.wpi.first.wpilibj2.command.SequentialCommandGroup;
import edu.wpi.first.wpilibj2.command.button.CommandXboxController;
import edu.wpi.first.wpilibj2.command.sysid.SysIdRoutine.Direction;

import frc.robot.generated.TunerConstants;
import frc.robot.subsystems.CommandSwerveDrivetrain;
import frc.robot.CoralShooter.CoralShooterContainer;

import edu.wpi.first.wpilibj.SerialPort;
import edu.wpi.first.wpilibj.SerialPort.Port;

public class RobotContainer {
     
    
    private double MaxSpeed = TunerConstants.kSpeedAt12Volts.in(MetersPerSecond); // kSpeedAt12Volts desired top speed
    private double MaxAngularRate = RotationsPerSecond.of(0.75).in(RadiansPerSecond); // 3/4 of a rotation per second max angular velocity

    /* Setting up bindings for necessary control of the swerve drive platform */
    private final SwerveRequest.FieldCentric drive = new SwerveRequest.FieldCentric()
            .withDeadband(MaxSpeed * 0.1).withRotationalDeadband(MaxAngularRate * 0.1) // Add a 10% deadband
            .withDriveRequestType(DriveRequestType.OpenLoopVoltage); // Use open-loop control for drive motors
    private final SwerveRequest.SwerveDriveBrake brake = new SwerveRequest.SwerveDriveBrake();
    private final SwerveRequest.PointWheelsAt point = new SwerveRequest.PointWheelsAt();

     private final TalonFX lClimber = new TalonFX(31); // left climber 
     private final TalonFX rClimber = new TalonFX(30); // right climber

    private final Telemetry logger = new Telemetry(MaxSpeed);

    public final CommandXboxController joystick = new CommandXboxController(0);

    public final CommandSwerveDrivetrain drivetrain = TunerConstants.createDrivetrain();
    private Command runauto1 = drivetrain.getAutoPath("Auto4");

    public double kSpeedDiv = 3.56;

    //public double kSpeedDiv = SmartDashboard.getNumber("Input Speed Div", 3.5); // 3.5 is our prefered speed for somewhat fast movements

    private SerialPort serial;

    public RobotContainer() {
        configureBindings();
    }

    private void configureBindings() {
        // Note that X is defined as forward according to WPILib convention,
        // and Y is defined as to the left according to WPILib convention.
        drivetrain.setDefaultCommand(
            // Drivetrain will execute this command periodically
            drivetrain.applyRequest(() ->
                drive.withVelocityX(-joystick.getLeftY() * MaxSpeed/kSpeedDiv) // Drive forward with negative Y (forward)
                    .withVelocityY(-joystick.getLeftX() * MaxSpeed/kSpeedDiv) // Drive left with negative X (left)
                    .withRotationalRate(-joystick.getRightX() * MaxAngularRate) // Drive counterclockwise with negative X (left)
            )
        );

        joystick.x().whileTrue(new InstantCommand(() -> {
            zeroPigeon();
            System.out.println("pigeon rezeroed");
        }));

        // Run SysId routines when holding back/start and X/Y.
        // Note that each routine should be run exactly once in a single log.
       // joystick.back().and(joystick.y()).whileTrue(drivetrain.sysIdDynamic(Direction.kForward));
       // joystick.back().and(joystick.x()).whileTrue(drivetrain.sysIdDynamic(Direction.kReverse));
        //joystick.start().and(joystick.y()).whileTrue(drivetrain.sysIdQuasistatic(Direction.kForward));
        //joystick.start().and(joystick.x()).whileTrue(drivetrain.sysIdQuasistatic(Direction.kReverse));

        // reset the field-centric heading on left bumper press

        drivetrain.registerTelemetry(logger::telemeterize);

  }
    
    

    public void zeroPigeon() {
    final Pigeon2 pigeon = drivetrain.getPigeon2();
    pigeon.reset();
  }

    public Command getAutonomousCommand(CoralShooterContainer launcher) {
        //return Commands.print("No autonomous command configured");
        //return runauto1;
        // SequentialCommandGroup auto_commands = new SequentialCommandGroup(
        //     Commands.print("Got Here!"),
        //     new PathPlannerAuto("Test")
        // );
        // return auto_commands;
        //return new PathPlannerAuto("Test");

        SequentialCommandGroup auto_commands = new SequentialCommandGroup(
          runauto1,
          new InstantCommand(() -> {launcher.shootCoral();}),
          new WaitCommand(1.0),
          new InstantCommand(() -> {launcher.stop();})
        );
        return auto_commands;
    }

    public void putSmartdashboardRobotContainer() {
        //SmartDashboard.putNumber("Input Speed Div", kSpeedDiv);
        //SmartDashboard.getNumber("Input Speed Div", kSpeedDiv);
        //SmartDashboard.setDefaultNumber("Input Speed Div", kSpeedDiv);
    }

    public void setLEDs() {
      serial = new SerialPort(9600, Port.kUSB1);

      serial.writeString("1\n");

      System.out.println("LED code called");
    }

    //public void quickSpeedControl() {

    //}
}

File: robot_container.py
"""
RobotContainer — ALL hardware initialization and ALL button bindings.

This centralizes controller mappings that were previously scattered across
Elevator, Climber, and CoralShooterContainer in the Java code.

Controls:
    Left stick:      Swerve translate (field-centric)
    Right stick X:   Swerve rotate
    X:               Reset Pigeon2 gyro
    Y:               Elevator up one level
    A:               Elevator down one level
    B:               Reset speed divider to 3.0
    Left bumper:     Decrease speed divider (faster)
    Right bumper:    Increase speed divider (slower)
    Right trigger:   Fire coral launcher (hold)
    D-pad up:        Climber up (hold)
    D-pad down:      Climber down (hold)
"""

from commands2 import InstantCommand, StartEndCommand
from commands2.button import CommandXboxController
from phoenix6.swerve.requests import FieldCentric

from constants import (
    DEFAULT_SPEED_DIVIDER,
    LAUNCHER_SPEED,
    MAX_ANGULAR_RATE,
    SPEED_AT_12V,
    create_drivetrain,
)
from subsystems.climber import Climber
from subsystems.coral_launcher import CoralLauncher
from subsystems.elevator import Elevator
from subsystems.limelight import Limelight
from subsystems.power_distribution_hub import PowerDistributionHub
from telemetry import Telemetry


class RobotContainer:
    def __init__(self):
        # Swerve drivetrain
        self.drivetrain = create_drivetrain()

        # Subsystems
        self.elevator = Elevator()
        self.climber = Climber()
        self.launcher = CoralLauncher()
        self.limelight = Limelight()
        self.pdh = PowerDistributionHub()

        # Telemetry
        self._max_speed = SPEED_AT_12V
        self._telemetry = Telemetry(self._max_speed)
        self.drivetrain.register_telemetry(self._telemetry.telemeterize)

        # Controller
        self.joystick = CommandXboxController(0)

        # Speed divider (captures by reference via self)
        self.speed_divider = 16.0  # starts at 16 until teleopInit resets to 3

        # Swerve requests
        self._drive = (
            FieldCentric()
            .with_deadband(self._max_speed * 0.1)
            .with_rotational_deadband(MAX_ANGULAR_RATE * 0.1)
        )

        # Wire up ALL bindings
        self._configure_bindings()

    def _configure_bindings(self):
        """Register all controller → command bindings once."""

        # --- Default command: field-centric swerve drive ---
        self.drivetrain.setDefaultCommand(
            self.drivetrain.apply_request(
                lambda: self._drive.with_velocity_x(
                    -self.joystick.getLeftY() * self._max_speed / self.speed_divider
                )
                .with_velocity_y(-self.joystick.getLeftX() * self._max_speed / self.speed_divider)
                .with_rotational_rate(-self.joystick.getRightX() * MAX_ANGULAR_RATE)
            )
        )

        # --- X: reset gyro ---
        self.joystick.x().onTrue(InstantCommand(lambda: self.drivetrain.get_pigeon2().reset()))

        # --- Y: elevator up one level ---
        self.joystick.y().onTrue(InstantCommand(self.elevator.go_up, self.elevator))

        # --- A: elevator down one level ---
        self.joystick.a().onTrue(InstantCommand(self.elevator.go_down, self.elevator))

        # --- D-pad up: climber up (hold) ---
        self.joystick.povUp().onTrue(InstantCommand(self.climber.extend, self.climber))
        self.joystick.povUp().onFalse(InstantCommand(self.climber.stop, self.climber))

        # --- D-pad down: climber down (hold) ---
        self.joystick.povDown().onTrue(InstantCommand(self.climber.retract, self.climber))
        self.joystick.povDown().onFalse(InstantCommand(self.climber.stop, self.climber))

        # --- Right trigger: fire coral launcher (hold) ---
        self.joystick.rightTrigger().whileTrue(
            StartEndCommand(
                lambda: self.launcher.launch(LAUNCHER_SPEED),
                self.launcher.stop,
                self.launcher,
            )
        )

        # --- Speed divider controls (registered once, not every loop) ---
        self.joystick.leftBumper().onTrue(InstantCommand(self._decrease_speed_divider))
        self.joystick.rightBumper().onTrue(InstantCommand(self._increase_speed_divider))
        self.joystick.b().onTrue(InstantCommand(self._reset_speed_divider))

    # -- Speed divider helpers (fix for quickJoystickSpeedChange bug) ---------

    def _decrease_speed_divider(self):
        """Make robot faster (decrease divider), clamped to 1.0."""
        self.speed_divider = max(1.0, self.speed_divider - 0.6)

    def _increase_speed_divider(self):
        """Make robot slower (increase divider), clamped to 12.0."""
        self.speed_divider = min(12.0, self.speed_divider + 0.6)

    def _reset_speed_divider(self):
        """Reset speed divider to default."""
        self.speed_divider = DEFAULT_SPEED_DIVIDER

    def reset_speed_for_teleop(self):
        """Called from Robot.teleopInit() to set teleop default speed."""
        self.speed_divider = DEFAULT_SPEED_DIVIDER
